File: src/main/python/AoC2015_14.py
#! /usr/bin/env python3
#
# Advent of Code 2015 Day 14
#

from __future__ import annotations

import sys
from collections import defaultdict
from dataclasses import dataclass
from typing import Self

from aoc.common import InputData
from aoc.common import SolutionBase

TEST = """\
Comet can fly 14 km/s for 10 seconds, but then must rest for 127 seconds.
Dancer can fly 16 km/s for 11 seconds, but then must rest for 162 seconds.
"""


@dataclass(frozen=True)
class Reindeer:
    name: str
    speed: int
    go: int
    stop: int

    @classmethod
    def from_input(cls, string: str) -> Self:
        splits = string.split()
        return cls(splits[0], int(splits[3]), int(splits[6]), int(splits[13]))

    def distance_reached(self, time: int) -> int:
        period_distance = self.speed * self.go
        period_time = self.go + self.stop
        periods = time // period_time
        left = time % period_time
        if left >= self.go:
            return periods * period_distance + period_distance
        return periods * period_distance + self.speed * left


Input = list[Reindeer]
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return [Reindeer.from_input(line) for line in input_data]

    def solve_1(self, reindeer: list[Reindeer], time: int) -> int:
        return max(r.distance_reached(time) for r in reindeer)

    def part_1(self, reindeer: Input) -> Output1:
        return self.solve_1(reindeer, 2503)

    def solve_2(self, reindeer: list[Reindeer], time: int) -> int:
        points = defaultdict[str, int](int)
        for i in range(time):
            distances = {r.name: r.distance_reached(i + 1) for r in reindeer}
            lead = max(distances.values())
            for r, d in distances.items():
                if d == lead:
                    points[r] += 1
        return max(points.values())

    def part_2(self, reindeer: Input) -> Output2:
        return self.solve_2(reindeer, 2503)

    def samples(self) -> None:
        reindeer = self.parse_input(TEST.splitlines())
        assert self.solve_1(reindeer, 1000) == 1120
        assert self.solve_2(reindeer, 1000) == 689


solution = Solution(2015, 14)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2015_14.java
import static com.github.pareronia.aoc.AssertUtils.assertTrue;
import static com.github.pareronia.aoc.StringOps.splitLines;
import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toMap;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.IntStream;

import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2015_14 extends SolutionBase<List<AoC2015_14.Reindeer>, Integer, Integer> {
    
    private AoC2015_14(final boolean debug) {
        super(debug);
    }

    public static final AoC2015_14 create() {
        return new AoC2015_14(false);
    }

    public static final AoC2015_14 createDebug() {
        return new AoC2015_14(true);
    }
    
    @Override
    protected List<Reindeer> parseInput(final List<String> inputs) {
        return inputs.stream().map(Reindeer::fromInput).collect(toList());
    }
    
    private int distance(final Reindeer reindeer, final int time) {
        final int periodDistance = reindeer.speed * reindeer.go;
        final int periodTime = reindeer.go + reindeer.stop;
        final int periods = time / periodTime;
        final int left = time % periodTime;
        if (left >= reindeer.go) {
            return periods * periodDistance + periodDistance;
        } else {
            return periods * periodDistance + reindeer.speed * left;
        }
    }

    private int solve1(final List<Reindeer> reindeer, final int time) {
        return reindeer.stream()
           .mapToInt(r -> distance(r, time))
           .max().orElseThrow();
    }

    private int solve2(final List<Reindeer> reindeer, final int time) {
        final Map<String, Integer> points = new HashMap<>();
        IntStream.rangeClosed(1, time).forEach(i -> {
            final Map<String, Integer> distances
                    = reindeer.stream().collect(toMap(r -> r.name, r -> distance(r, i)));
            final int lead = distances.values().stream().mapToInt(Integer::intValue).max().orElseThrow();
            distances.entrySet().stream()
                .filter(e -> e.getValue() == lead)
                .forEach(e -> points.merge(e.getKey(), 1, Integer::sum));
        });
        return points.values().stream().mapToInt(Integer::intValue).max().orElseThrow();
    }
    
    @Override
    public Integer solvePart1(final List<Reindeer> reindeer) {
        return solve1(reindeer, 2503);
    }
    
    @Override
    public Integer solvePart2(final List<Reindeer> reindeer) {
        return solve2(reindeer, 2503);
    }

    public static void main(final String[] args) throws Exception {
        final AoC2015_14 test = AoC2015_14.createDebug();
        final List<Reindeer> input = test.parseInput(TEST);
        assert test.solve1(input, 1000) == 1120;
        assert test.solve2(input, 1000) == 689;

        AoC2015_14.create().run();
    }
    
    private static final List<String> TEST = splitLines(
        "Comet can fly 14 km/s for 10 seconds, but then must rest for 127 seconds.\r\n" +
        "Dancer can fly 16 km/s for 11 seconds, but then must rest for 162 seconds."
    );
    
    record Reindeer(String name, int speed, int go, int stop) {
        
        private static final String REGEXP
                = "([A-Za-z]+) can fly ([0-9]+) km/s for ([0-9]+) seconds," +
                        " but then must rest for ([0-9]+) seconds\\.";
        private static final Pattern pattern = Pattern.compile(REGEXP);
        
        public static Reindeer fromInput(final String input) {
           final Matcher m = pattern.matcher(input);
           assertTrue(m.matches(), () -> "No match found");
           return new Reindeer(m.group(1),
                   Integer.parseInt(m.group(2)),
                   Integer.parseInt(m.group(3)),
                   Integer.parseInt(m.group(4)));
        }
    }
}
